waku/v2/protocol/rln/group_manager/dynamic/dynamic.go

```diff
@@ -278,17 +297,28 @@ func (gm *DynamicGroupManager) persistCredentials() error {
 func (gm *DynamicGroupManager) InsertMembers(toInsert *om.OrderedMap) error {
 	for pair := toInsert.Oldest(); pair != nil; pair = pair.Next() {
 		events := pair.Value.([]*contracts.RLNMemberRegistered) // TODO: should these be sortered by index? we assume all members arrive in order
+		var idCommitments []rln.IDCommitment
+		var oldestIndexInBlock *big.Int
 		for _, evt := range events {
-			pubkey := rln.Bytes32(evt.Pubkey.Bytes())
-			// TODO: should we track indexes to identify missing?
-			err := gm.rln.InsertMember(pubkey)
-			if err != nil {
-				gm.log.Error("inserting member into merkletree", zap.Error(err))
-				return err
+			if oldestIndexInBlock == nil {
+				oldestIndexInBlock = evt.Index
 			}
+			idCommitments = append(idCommitments, rln.Bytes32(evt.Pubkey.Bytes()))
+		}
+
+		if len(idCommitments) == 0 {
+			continue
+		}
+
+		// TODO: should we track indexes to identify missing?
+		startIndex := rln.MembershipIndex(uint(oldestIndexInBlock.Int64()))
+		err := gm.rln.InsertMembers(startIndex, idCommitments)
```

Comment:
Does the RLN API take care of a scenario where few members are already inserted and not throw error? If not, this has to be handled at this level.

e.g: While inserting members, let's say member insertion succeeded whereas before metadata got updated, node went down. Next time when node comes up, it will try to insert some set of members which are already inserted and probably a new set if they are part of a later block. In this case, atomicOp should not fail. If it fails, then we may have to handle it at this layer(identify if members are already present and not insert them again).

Reply:
> Does the RLN API take care of a scenario where few members are already inserted and not throw error? I

The RLN API handles it by acting as an upsert, overwriting whatever is stored at some index in the merkle tree, so atomicOp should not fail as it will just replace the same members at the same position with the same values